### PR TITLE
Use relative transformations

### PR DESCRIPTION
**Pull request by helgee**
This addresses #46 by not basing all epochs on TAI anymore and only using relative transformations. It also stops abusing the type system by not using values-as-parameters (see e.g. [here](https://docs.julialang.org/en/v1/manual/performance-tips/#The-dangers-of-abusing-multiple-dispatch-(aka,-more-on-types-with-values-as-parameters)-1)) and will hopefully improve latency.

### CODE REVIEW

**Comment by helgee** [comment]
There are missing/outdated docs but this is feature-complete now. @giordano and @bgodard if you could have a look I'd be much obliged.

File: test/Project.toml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+[deps]
```
Comment (giordano) [inline test/Project.toml]: Is this file used?

File: src/accessors.jl [hunk 2]
```diff
@@ -6,27 +6,20 @@ import ..AstroDates: DateTime, year, month, day,
...
-    d′ = sum - o′
-    Δo = ep.offset - o′
-    Δd = ep.ts_offset - d′
-    residual = Δo + Δd
-
-    carry = floor(Int64, sum)
-    offset2000B = (sum - carry) + residual
-    offset2000A = ep.epoch + carry + Int64(43200)
+    carry = floor(Int64, ep.fraction)
+    offset2000B = ep.fraction - carry
```
Comment (giordano) [inline src/accessors.jl]: I think this is similar to
```julia
offset2000B, carry = modf(ep.fraction)
```
but I guess you really want to have `carry` as an integer?

File: src/offsets.jl [hunk 3]
```diff
@@ -176,27 +346,18 @@ function tai_offset(::BarycentricDynamicalTime, ep, elong, u, v)
 
     # TDB-TT in seconds.
     w = wt + wf + wj
-
-    tai_offset(TT, ep) + w
+    return -w
```
Comment (giordano) [inline src/offsets.jl]: Just do
```julia
return -(wt + wf + wj)
```
?  Or you prefer to show the intermediate step?

File: test/Project.toml (hunk 1 above)
Comment (helgee) [reply test/Project.toml]: On Julia >= 1.3 it is.

File: src/offsets.jl (hunk 3 above)
Comment (helgee) [reply src/offsets.jl]: During development it is nice to have the lines separated so one can quickly prefix `@show` for debugging. But for the final result I do not really care.

File: src/accessors.jl (hunk 2 above)
Comment (helgee) [reply src/accessors.jl]: I think this whole code path should be unnecessary. `ep.fraction < 1.0` should hold by definition. Will check 👍

**Comment by helgee** [comment]
@bgodard and I are having an offline discussion about the inherent trade-offs of values-as-parameters.

E.g. `rescale`

```julia
struct Epoch{S<:TimeScale}
    scale::S
    second::Int64
    fraction::Float64
end
```
vs. `master`
```julia
struct Epoch{S}
    second::Int64
    fraction::Float64
    function Epoch{S}(second, fraction)
        new{S::TimeScale}(second, fraction)
    end
end
```
@giordano in which camp do you fall?

**Comment by bgodard** [comment]
Just to be clear my preference would be to use a Type as parameter. But to remove the scale value in the Epoch structure since the timescale can be deduced from the type, hence no value is needed.

eg:
```julia
abstract Timescale end
abstract TDBScale <: Timscale end
abstract TTScale <: TimeScale end

struct Epoch{S<:TimeScale}
    second::Int64
    fraction::Float64
end
```

**Comment by helgee** [comment]
That's how things used to be a long time ago: https://github.com/JuliaAstro/AstroTime.jl/blob/85ba00c5e6da6567a9f557722372d2a6b070d5b4/src/AstronomicalTime.jl

Of course, Julia was quite different back then. But I remember that (ab-)using abstract types in such a way led to all sorts of funny problems.

I made the following experiment:
```julia
using BenchmarkTools

abstract type TimeScale end
abstract type TDBScale <: TimeScale end
abstract type TTScale <: TimeScale end

struct ConcreteTDBScale <: TimeScale end
const TDB = ConcreteTDBScale()
struct ConcreteTTScale <: TimeScale end
const TT = ConcreteTTScale()

const k = 1.657e-3
const eb = 1.671e-2
const m₀ = 6.239996
const m₁ = 1.99096871e-7

struct AEpoch{S<:TimeScale}
    second::Int64
    fraction::Float64
end

function getoffset(::Type{TTScale}, ::Type{TDBScale}, ep::AEpoch{TTScale})
    tt = ep.fraction + ep.second
    g = m₀ + m₁ * tt
    return k * sin(g + eb * sin(g))
end

struct CEpoch{S<:TimeScale}
    scale::S
    second::Int64
    fraction::Float64
end

function getoffset(::ConcreteTTScale, ::ConcreteTDBScale, ep::CEpoch{ConcreteTTScale})
    tt = ep.fraction + ep.second
    g = m₀ + m₁ * tt
    return k * sin(g + eb * sin(g))
end

aep = AEpoch{TTScale}(0, 0.0)
@show sizeof(aep)
@benchmark getoffset($TTScale, $TDBScale, $aep)

cep = CEpoch(ConcreteTTScale(), 0, 0.0)
@show sizeof(cep)
@benchmark getoffset($TT, $TDB, $cep)
```

This is the result:

```
sizeof(aep) = 16

BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     0.049 ns (0.00% GC)
  median time:      0.057 ns (0.00% GC)
  mean time:        0.062 ns (0.00% GC)
  maximum time:     33.603 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000

sizeof(cep) = 16

BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     0.047 ns (0.00% GC)
  median time:      0.057 ns (0.00% GC)
  mean time:        0.057 ns (0.00% GC)
  maximum time:     0.287 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000
```

The performance is virtually identical and even the size of the struct in memory is the same. I would argue though that the approach with an explicit `scale` field gives cleaner code (not significantly in this toy example but in the library, e.g. far fewer `Type{}` wrappers) and is less arcane and much easier to understand for Julia beginners.

**Comment by giordano** [comment]
Subnanoseconds benchmarks are usually indication that constant-propagation kicked in, making benchmarks useless.  These numbers should be a bit more realistic, as `Ref` prevents constant-folding:
```julia
julia> @benchmark getoffset($(Ref(TTScale))[], $(Ref(TDBScale))[], $(Ref(aep))[])
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     52.585 ns (0.00% GC)
  median time:      53.596 ns (0.00% GC)
  mean time:        54.543 ns (0.00% GC)
  maximum time:     163.054 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     986

julia> @benchmark getoffset($(Ref(TT))[], $(Ref(TDB))[], $(Ref(cep))[])
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     37.165 ns (0.00% GC)
  median time:      37.188 ns (0.00% GC)
  mean time:        37.741 ns (0.00% GC)
  maximum time:     107.502 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     992
```

**Comment by helgee** [comment]
Thanks for pointing this out @giordano. My brain was apparently still asleep this morning 😜

**Comment by helgee** [comment]
@bgodard I have revived the parametrization of `Epoch` and added some basic tests based on [Measurements.jl](https://github.com/JuliaPhysics/Measurements.jl), see here: https://github.com/JuliaAstro/AstroTime.jl/pull/48/commits/941564545382d6379a5f4efae3d8953156ad7e0b#diff-4d59095798497247b433c8a8882d6854R210

**Comment by giordano** [comment]
Regarding the general stylistic question, I don't have a strong opinion, but I agree that having a concrete field might be a little bit cleaner.

**Comment by helgee** [comment]
If all goes well (Travis seems to be in a bad mood right now), this should be ready to merge now 🎉

**Comment by helgee** [comment]
Here is a preview of the reworked docs: https://juliaastro.github.io/AstroTime.jl/previews/PR48/

**Comment by helgee** [comment]
I am going to merge this tomorrow if there are no objections.

**Comment by bgodard** [comment]
Concerning the storage of the timescale in the Epoch struct what we observe is that the timescale variable has size 0 so it does not use any memory which is good (I was worried it would reserve some space for object identity check like in C++ but julia doc says that empty struct are singleton so all instances are identical), but I wonder what makes it faster in your example when you add the timescale record. 

I still find more confusing as a user to have a kind of duplicate information about timescale (fully determined by type and also by instance even though julia optimizes out the instance. 

The other approach seems so common in julia standard library:
eg in Base. number.jl
`convert(::Type{T}, x::T)      where {T<:Number} = x`
Of course the other option is not possible here since porimitive type do not have fields.

Do you have any example of other julia library storing an instance of an empty struct?

**Comment by bgodard** [comment]
Also Units.jl use the other style, but it is deprecated. 

In the more recent Unitful, types.jl:
```julia
struct Unit{U,D}
        tens::Int
        power::Rational{Int}
    end
```
No instance of types U and D are stored.

**Comment by bgodard** [comment]
Correction in Unitful U and D are not types! Sorry about this.

**Comment by helgee** [comment]
> I still find more confusing as a user to have a kind of duplicate information about timescale (fully determined by type and also by instance even though julia optimizes out the instance.

I would argue that this is an implementation detail and should actually be invisible to the user in day-to-day use.

> Do you have any example of other julia library storing an instance of an empty struct?

DifferentialEquations.jl is a good example. See e.g. [here](https://github.com/JuliaDiffEq/DiffEqBase.jl/blob/master/src/problems/ode_problems.jl) (the code is pretty dense though). There is a `problem_type` field and all the problem types are empty structs.

> Correction in Unitful U and D are not types! Sorry about this.

No worries 👍

**Pull request merged**